### PR TITLE
feat: integrate million js automatic mode

### PR DESCRIPTION
**Pull request by kunal00000**
Resolves https://github.com/aidenybai/million/issues/556
/claim https://github.com/aidenybai/million/issues/556

the below demo shows suggestions and improvements given by million js in automatic mode.
@jhechtf this now works as expected.

https://github.com/bautistaaa/typehero/assets/92316166/414fe507-733f-4c8c-9253-19b4c48306e3

### CODE REVIEW

**Comment by jhechtf** [comment]
@kunal00000 I have pulled down your code and done a clean/prune of my local to ensure it's not user error, but the code still breaks for me.

Screenshot 2023-08-25 at 2 36 35 PM

To further ensure that it wasn't just my machine I authorized the code to go to a preview env and that also has some build errors as well -- how are you getting this to work locally?

@Hacksore or @bautistaaa either of you able to get this working to approve it?

**Comment by kunal00000** [comment]
The error shown in the image is because of million js, it is basically suggesting to improve this chunk of code.

To ignore this
- We can mute these help messages
- ignore component 
```
// million-ignore
function App() {
  return ...
}
```

**Comment by aidenybai** [comment]
@kunal00000 @jhechtf The correct config is `auto: { rsc: true }` since this is a RSC project

File: apps/admin/next.config.js [hunk 1]
```diff
@@ -1,3 +1,9 @@
+import million from 'million/compiler';
+
+const millionConfig = {
+  auto: true,
```
Comment (aidenybai) [inline apps/admin/next.config.js:4]:
```suggestion
  auto: { rsc: true },
```

File: apps/og-image/next.config.js [hunk 2]
```diff
@@ -1,3 +1,9 @@
+import million from 'million/compiler';
+
+const millionConfig = {
+  auto: true,
```
Comment (aidenybai) [inline apps/og-image/next.config.js:4]: I would not recommend using Million.js here since it's a OG image gen

File: apps/web/next.config.js [hunk 3]
```diff
@@ -1,3 +1,9 @@
+import million from 'million/compiler';
+
+const millionConfig = {
+  auto: true,
```
Comment (aidenybai) [inline apps/web/next.config.js:4]:
```suggestion
  auto: { rsc: true },
```

File: apps/admin/next.config.js (hunk 1 above)
Comment (kunal00000) [reply apps/admin/next.config.js]: @aidenybai what is RSC project btw?

**Comment by kunal00000** [comment]
@aidenybai pushed the changes according to reviews.

File: apps/admin/next.config.js (hunk 1 above)
Comment (bautistaaa) [reply apps/admin/next.config.js]: nextjs app router uses React Server Components(RSC)

Comment (kunal00000) [reply apps/admin/next.config.js]: Ah okay, Got it 👍🏻.

**Comment by Hacksore** [comment]
@aidenybai how stable is this for RSC?

**Comment by aidenybai** [comment]
@Hacksore afaik at this time, 2 companies are using million in rsc and 3 are using pages router

**Comment by Hacksore** [comment]
> @Hacksore afaik at this time, 2 companies are using million in rsc and 3 are using pages router, both in prod

😅 - we are bleeding edge™

File: pnpm-lock.yaml [hunk 4]
```diff
@@ -1,9 +1,5 @@
 lockfileVersion: '6.0'
 
-settings:
-  autoInstallPeers: true
-  excludeLinksFromLockfile: false
```
Comment (QuiiBz) [inline pnpm-lock.yaml:5]: Any reason why these settings have been removed?

**Review by QuiiBz — CHANGES_REQUESTED** [review]
Tried to run it locally and we have some warnings and errors:

Screenshot 2023-08-27 at 13 34 11

Screenshot 2023-08-27 at 13 35 15

Would you like to take a look @kunal00000? Also pinging @aidenybai as I'm not sure if there errors are normal or not.

File: pnpm-lock.yaml (hunk 4 above)
Comment (kunal00000) [reply pnpm-lock.yaml:5]: this is not removed instead it is shifted to last of the file, this happened on running `pnpm install`

**Comment by kunal00000** [comment]
@QuiiBz I think these are normal warnings, suggesting to improve the project's performance.
curious to know from @aidenybai, he will be able to give better answer for this.

**Comment by aidenybai** [comment]
i just pushed a fix? for this @kunal00000 @QuiiBz, try installing latest

**Comment by QuiiBz** [comment]
@aidenybai it fixed the errors but there is a new one. Happy to jump on a call to debug this further:

Screenshot 2023-08-28 at 07 47 58

**Comment by bautistaaa** [comment]
any updates here? would like to get this in at some point

**Comment by kunal00000** [comment]
@bautistaaa should I update branch and million dependency if you are open to accept this?

**Comment by Hacksore** [comment]
Yes @kunal00000 update the deps and branch and let's see if we run into any more issues before considering merging this.

**Comment by kunal00000** [comment]
@Hacksore updated the branch and deps. Please check if it leads to issues.

**Comment by QuiiBz** [comment]
Just tried and here are the results:
- Explore, Waitlist, and Profile pages works fine
- Landing and Track pages both fail with this error:
Screenshot 2023-09-01 at 16 31 30

- Challenge page fails with this error:
Screenshot 2023-09-01 at 16 33 37

Commenting the line that has the errors resolves the issues.

**Comment by kunal00000** [comment]
hi @QuiiBz, if these are not major issues they can be ignored using `// million-ignore` at the top of component (if that works).

**Comment by rishi-raj-jain** [comment]
have encountered some of the issues while contributing to lenster (https://github.com/lensterxyz/lenster/pull/3632) as it changes the scope of rendering.

https://github.com/lensterxyz/lenster/pull/3632#issuecomment-1703813815 is the reason the variables are breaking (the FollowerNotification.tsx breaks in lenster) and all of it just starts working if I add million-ignore. The optimized block, fails to load the props from the component which it is destructured from and rather look for alike prop variable in the createPortal function.

**Comment by aidenybai** [comment]
@QuiiBz sorry! i did not receive notifications for this for some reason. i may have pushed a fix for this recently, can you check if latest fixes the bug?

**Comment by QuiiBz** [comment]
Updated from beta 11 to beta 12, everything seems to work fine. I only had to ignore our `SplitEditor` component, because it would lose the Monaco editor focus after editing a char (strange bug): https://github.com/typehero/typehero/pull/430/files#diff-e37b517b7fd5dca0d50564f5f70eb44744a5d16a4dd2481503dd8fd305497832

**Review by QuiiBz — APPROVED** [review]

**Pull request merged**